src/pulling_gauge.rs

```diff
@@ -0,0 +1,75 @@
+use std::{collections::HashMap, fmt, sync::Arc};
+
+use crate::{
+    core::Collector,
+    proto::{Gauge, Metric, MetricFamily, MetricType},
+};
+use protobuf::RepeatedField;
+
+/// A prometheus gauge that reads the value from a provided function on every
+/// collect run. Used to expose metrics that are better to be polled as opposed
+/// to continuously set.
+#[derive(Clone)]
+pub struct PullingGauge {
+    desc: crate::core::Desc,
+    value: Arc<Box<dyn Fn() -> f64 + Send + Sync>>,
+}
+
+impl fmt::Debug for PullingGauge {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("PullingGauge")
+            .field("desc", &self.desc)
+            .field("value", &"<opaque>")
+            .finish()
+    }
+}
+
+impl PullingGauge {
+    /// Create a new PullingGauge.
+    pub fn new<S1: Into<String>, S2: Into<String>>(
+        fq_name: S1,
+        help: S2,
+        value: Box<dyn Fn() -> f64 + Send + Sync>,
```

Comment:
A general question about the function here: does this assume that you have somewhere else some global mutable state? Or does this work with non-mutable captures too? (I'm mostly concerned about the lifetime of the underlying state)

Reply:
It does work with non-mutable captures too but typically we have some global state that we want to monitor. We typically use this to either
- fetch a system/runtime metric (e.g. std::thread::available_parallelism or tokio runtime metrics)
- report on the length of a queue by e.g. passing a `Arc<RwLock<>>` or an `ArcSwap`